src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs

```diff
@@ -0,0 +1,337 @@
+                        releaseList.AddRange(runningPipelinesForBuild.Value);
+                        await runningPipelines.AddOrUpdateAsync(tx, buildId, releaseList, (key, oldValue) => releaseList);
+                    }
+                    else
+                    {
+                        await runningPipelines.AddAsync(tx, buildId, releaseList);
+                    }
+                    await tx.CommitAsync();
+                }
+            }
+            await ProcessFinishedReleasesAsync(cancellationToken);
+        }
+
+        /// <summary>
+        /// Check for finished releases and add the corresponding build channel association to the DB.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        // Check every 10 minutes for any finished release Pipelines
+        [CronSchedule("0 0/10 * 1/1 * ? *", TimeZones.UTC)]
+        public async Task ProcessFinishedReleasesAsync(CancellationToken cancellationToken)
+        {
+            var runningPipelines =
+                        await StateManager.GetOrAddAsync<IReliableDictionary<int, IList<ReleasePipelineStatusItem>>>(RunningPipelineDictionaryName);
+            try
+            {
+                HashSet<BuildChannel> buildChannelsToAdd = new HashSet<BuildChannel>(new BuildChannelComparer());
+                using (ITransaction tx = StateManager.CreateTransaction())
+                {
+                    var runningPipelinesEnumerable = await runningPipelines.CreateEnumerableAsync(tx, EnumerationMode.Unordered);
+                    var asyncEnumerator = runningPipelinesEnumerable.GetAsyncEnumerator();
+                    while (await asyncEnumerator.MoveNextAsync(cancellationToken))
+                    {
+                        int buildId = asyncEnumerator.Current.Key;
+                        ConditionalValue<IList<ReleasePipelineStatusItem>> maybeReleaseStatuses = await runningPipelines.TryGetValueAsync(tx, buildId);
+                        if (maybeReleaseStatuses.HasValue)
+                        {
+                            bool allFinished = true;
+                            IList<ReleasePipelineStatusItem> releaseStatuses = maybeReleaseStatuses.Value;
+
+                            foreach (ReleasePipelineStatusItem releaseStatus in releaseStatuses)
+                            {
+                                int releaseId = releaseStatus.ReleaseId;
+                                AzureDevOpsClient azdoClient = await GetAzureDevOpsClientForAccount(releaseStatus.PipelineOrganization);
+                                AzureDevOpsRelease release = await azdoClient.GetReleaseAsync(releaseStatus.PipelineOrganization, releaseStatus.PipelineProject, releaseStatus.ReleaseId);
+                                string currentStatus = release.Environments[0].Status;
```

Comment:
>[0] [](start = 75, length = 3)

Should we check all the environments? Or log something if there is not exactly one?

Reply:
I'll check the others. It's the correct thing to do